File: Local Labor Market Data/Model/Report/CESReport.swift
//
//  CESReport.swift
//  Local Labor Market Data
//
//  Created by Nidhi Chawla on 7/30/18.
//  Copyright © 2018 Department of Labor. All rights reserved.
//

import Foundation

class CESReport {
    enum DataTypeCode: String {
        case allEmployees = "01"
        case avgWeeklyHours_AllEmployees = "02"
        case avgHourlyEarnings_AllEmployees = "03"
        case production_NonsupervisoryEmployees = "06"
        case avgWeeklyHours_ProductionEmployees = "07"
        case avgHourlyEarnings_ProductionEmployees = "08"
        case avgWeeklyEarnings_AllEmployees = "11"
        case allEmployees_3MonthAvgChange = "26"
        case averageWeeklyEarlnings_ProductionEmployees = "30"
    }

//    class func getEmploymentLevelSeriesId(forArea area: Area,
//                                           adjustment: SeasonalAdjustment) -> SeriesId? {
//        return getSeriesId(forArea: area, measureCode: .UnemploymentRate, adjustment: adjustment)
//    }
    
    class func getSeriesId(forArea area: Area, industryCode: String = "00000000",
                           dataTypeCode: DataTypeCode,
                           adjustment: SeasonalAdjustment) -> SeriesId? {
        if area is National {
            return getNationalSeriesId(dataTypeCode: dataTypeCode, adjustment: adjustment)
        }
        else {
            return getLocalAreaSeriesId(forArea: area, industryCode: industryCode, dataTypeCode: dataTypeCode, adjustment: adjustment)
        }
    }
    
    class func getNationalSeriesId(industryCode: String = "00000000",
                                   dataTypeCode: DataTypeCode,
                                   adjustment: SeasonalAdjustment) -> SeriesId? {
        
        let seriesId = "CE" + adjustment.rawValue  + industryCode + dataTypeCode.rawValue
        return seriesId
    }
    
    class func getLocalAreaSeriesId(forArea area: Area, industryCode: String = "00000000",
                                   dataTypeCode: DataTypeCode,
                                   adjustment: SeasonalAdjustment) -> SeriesId? {
        
        guard let code = area.code else { return nil}
       var areaCode: String = ""
        if let metro = area as? Metro, let stateCode = metro.stateCode {
            areaCode = stateCode
        }
        
        areaCode.append(code)
        areaCode = areaCode.rightPadding(toLength: 7, withPad: "0")
        let seriesId = "SM" + adjustment.rawValue  + areaCode + industryCode + dataTypeCode.rawValue
        return seriesId
    }
}

File: app/src/main/java/blsapp/dol/gov/blslocaldata/model/reports/CESReport.kt
package blsapp.dol.gov.blslocaldata.model.reports

import blsapp.dol.gov.blslocaldata.db.entity.AreaEntity
import blsapp.dol.gov.blslocaldata.db.entity.MetroEntity
import blsapp.dol.gov.blslocaldata.db.entity.NationalEntity

class CESReport {
    enum class DataTypeCode(val code: String) {
        ALLEMPLOYEES("01"),
        AVGWEEKLYHOURS_ALLEMPLOYEES("02"),
        AVGHOURLYEARNINGS_ALLEMPLOYEES("03"),
        PRODUCTION_NONSUPERVISORYEMPLOYEES("06"),
        AVGWEEKLYHOURS_PRODUCTIONEMPLOYEES("07"),
        AVGHOURLYEARNINGS_PRODUCTIONEMPLOYEES("08"),
        AVGWEEKLYEARNINGS_ALLEMPLOYEES("11"),
        ALLEMPLOYEES_3MONTHAVGCHANGE("26"),
        AVERAGEWEEKLYEARLNINGS_PRODUCTIONEMPLOYEES("30")
    }

    companion object {
        fun getSeriesId(area: AreaEntity, industryCode: String = "00000000",
                    dataTypeCode: DataTypeCode,
                    adjustment: SeasonalAdjustment): SeriesId {

            when (area) {
                is NationalEntity -> {
                    return getNationalSeriesId(dataTypeCode = dataTypeCode, adjustment = adjustment)
                }
            }

            return getLocalAreaSeriesId(area, dataTypeCode = dataTypeCode, adjustment = adjustment)
        }

        fun getNationalSeriesId(industryCode: String = "00000000",
                        dataTypeCode: DataTypeCode,
                        adjustment: SeasonalAdjustment): SeriesId {

            val seriesId = "CE" + adjustment.value  + industryCode + dataTypeCode.code
            return seriesId
        }

        private fun getLocalAreaSeriesId(area: AreaEntity, industryCode: String = "00000000",
                                 dataTypeCode: DataTypeCode,
                                    adjustment: SeasonalAdjustment) : SeriesId {

            val code = area.code
            var areaCode = ""

            if (area is MetroEntity) {
                areaCode = area.stateCode
            }

            areaCode += code
            areaCode = areaCode.padEnd(7, '0')
            val seriesId = "SM" + adjustment.value  + areaCode + industryCode + dataTypeCode.code
            return seriesId
        }
    }
}